### PR TITLE
fixes computation of service default parameters while registering tokens

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- fixes computation of service default parameters while registering tokens

## Why are we making these changes?

Users should be able to create tokens where defaults for required parameters are provided by the profile.

### CODE REVIEW

File: waiter/src/waiter/token.clj [hunk 1]
```diff
@@ -434,7 +434,8 @@
       (when (not-empty unknown-keys)
         (throw (ex-info (str "Unsupported key(s) in token: " (str (vec unknown-keys)))
                         {:status http-400-bad-request :token token :log-level :warn}))))
-    (let [service-parameter-with-token-defaults (attach-token-defaults-fn new-service-parameter-template)]
+    (let [service-parameter-with-token-defaults (attach-service-defaults-fn new-service-parameter-template)
+          missing-parameters (->> sd/service-required-keys (remove #(contains? new-service-parameter-template %1)) seq)]
```
Comment (DaoWen) [inline waiter/src/waiter/token.clj:438]: Why is `missing-parameters` defined in terms of `new-service-parameter-template` and not `service-parameter-with-token-defaults`?

File: waiter/src/waiter/token.clj [hunk 2]
```diff
@@ -434,7 +434,8 @@
       (when (not-empty unknown-keys)
         (throw (ex-info (str "Unsupported key(s) in token: " (str (vec unknown-keys)))
                         {:status http-400-bad-request :token token :log-level :warn}))))
-    (let [service-parameter-with-token-defaults (attach-token-defaults-fn new-service-parameter-template)]
+    (let [service-parameter-with-token-defaults (attach-service-defaults-fn new-service-parameter-template)
```
Comment (DaoWen) [inline waiter/src/waiter/token.clj:437]: Did we audit all other usages of `attach-token-defaults` to make sure this mistake isn't repeated elsewere?

**Review by DaoWen — CHANGES_REQUESTED** [review]

File: waiter/src/waiter/token.clj (hunk 2 above)
Comment (DaoWen) [reply waiter/src/waiter/token.clj:437]: Also, this should be renamed <i>service-parameter-with-<b>service</b>-defaults</i>.

Comment (shamsimam) [reply waiter/src/waiter/token.clj]: The [previous PR which wanted to fix the bug](https://github.com/twosigma/waiter/pull/1197/files) introduced `attach-token-defaults` into the file. It has now been replaced with uses of `attach-service-defaults` in this file.

Also, renamed variable to `service-parameter-with-service-defaults`.

File: waiter/src/waiter/token.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Oops, copy-paste error. Fixed and added assertions in tests to validate the reported missing parameters.

**Review by DaoWen — APPROVED** [review]

File: waiter/src/waiter/token.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: > Did we audit all other usages of attach-token-defaults to make sure this mistake isn't repeated elsewhere?

I audited all uses in this file where we introduced the function as part of a [recent PR](https://github.com/twosigma/waiter/pull/1197/files).

I also checked the other uses. In all but one usage, the function is passed around as a pair to `attach-service-defaults`. The only single pass of `attach-token-defaults` is to compute token GC time which relies on token only parameters: `fallback-period-secs` and `stale-timeout-mins` in `waiter.service-description/source-tokens->gc-time-secs`.

**Review by DaoWen — APPROVED** [review]

**Pull request merged**